### PR TITLE
libpulseaudio: split packages properly

### PR DESCRIPTION
**Pull request by tomty89**

### CODE REVIEW

**Comment by fornwall** [comment]
Thanks a lot!

As I'm a bit green on pulseaudio, could @ackalker, @Grimler91, @its-pointless (who have worked on pulseaudio in Termux before) or anyone else review this?

**Comment by Grimler91** [comment]
Haven't tested this yet, but I guess TERMUX_PKG_BREAKS is needed for both libpulseaudio and pulseaudio, since files are moved between the packages. 

See https://www.debian.org/doc/debian-policy/#s-conflicts and https://www.debian.org/doc/debian-policy/#s-replaces.

**Comment by Grimler91** [comment]
I thought we added TERMUX_SUBPKG_BREAKS before, but apparently not, need to fix that.

**Comment by tomty89** [comment]
@Grimler91 I think when libpulseaudio "breaks" pulseaudio, old pulseaudio is ensured to be deconfigured before new libpulseaudio is installed; while new pulseaudio is ensured to be installed after new libpulseaudio because of the dependency, so no breaks needed in new pulseaudio?

**Comment by Grimler91** [comment]
@tomty89 I think you would run into trouble if you manually update pulseaudio but not libpulseaudio, as pulseaudio then have files conflicting with libpulseaudio. 

So if the user runs `apt update && apt install pulseaudio` it fails, but `apt update && apt install libpulseaudio` would upgrade both packages and succeed. 
Most users probably wouldn't do this cherry-picking upgrade but it should be considered.

**Comment by tomty89** [comment]
@Grimler91 I think that should be dealt with versioned dependency? I was just not sure if Termux has a general policy on that.

**Comment by tomty89** [comment]
@Grimler91 Also is it even appropriate for a user to cherry-pick upgrade like that? I mean, one should do `apt update && apt install libpulseaudio pulseaudio`, no? I don't even expect `apt update && apt install libpulseaudio` to work.

**Comment by Grimler91** [comment]
@tomty89 I'm not entirely sure but I don't think versioned dependency would solve for example the cherry-pick upgrade case. Probably best to try it out. 

Appropriate, probably not, but maybe a user has limited bandwidth and does this, for example.

**Comment by tomty89** [comment]
@Grimler91 what I meant was since they are split packages, a user isn't supposed to only upgrade pulseaudio but not libpulseaudio. In my opinion:

1. `apt update && apt install libpulseaudio pulseaudio` should be guaranteed to work, as is `apt upgrade`.
2. `apt update && apt install libpulseaudio` *should fail* because `libpulseaudio` wasn't supposed to pull `pulseaudio` anyway; so if a user *really just* want to upgrade libpulseaudio, pulseaudio *should be ditched and it is his own responsibility*
3. *ideally* `apt update && apt install pulseaudio` should be guaranteed to work because pulseaudio pulls libpulseaudio, and I *assume* with versioned dependency it would work; however it doesn't seem to me that versioned dependency between split packages is a policy in Termux (yet), so I don't feel like doing it in just pulseaudio, especially when it is really just for some corner case (which can be worked around with way 1)

My background is mostly Arch so I am not familar with the debian way/tools (Seriously I don't really get it even after reading the documentation). I think I am going to leave it to @fornwall

**Comment by fornwall** [comment]
@tomty89 I agree with (3) at the moment!

In the future we can perhaps improve dependency versioning with something automatic (like always requiring at least the version of the dependency that was used at build time), but for now I don't think we need to spend the time versioning all dependencies manually.

**Comment by Grimler91** [comment]
@tomty89 Correct me if I'm wrong but you're talking about API changes in libpulseaudio.so that breaks pulseaudio (for example)? 
Updating only pulseaudio would then of course give a none-working package since it's linked against a different library. But the actual command `apt update && apt install pulseaudio` would succeed.

If files are moved between the packages and TERMUX_{SUB,}PKG_BREAKS isn't set however, then the command `apt update && apt install pulseaudio` would fail. 

I'm mostly interested in avoiding apt errors here.

**Comment by tomty89** [comment]
> Updating only pulseaudio would then of course give a none-working package since it's linked against a different library. But the actual command apt update && apt install pulseaudio would succeed.

Why do we want it to succeed then?

> If files are moved between the packages and TERMUX_{SUB,}PKG_BREAKS isn't set however, then the command apt update && apt install pulseaudio would fail.

Why should we care when we actually want it to fail even when files are *not* moved between them?

**Comment by tomty89** [comment]
> Updating only pulseaudio would then of course give a none-working package since it's linked against a different library. But the actual command apt update && apt install pulseaudio would succeed.

I am not sure what's your point in this. Yes it would succeed (when no files are moved I suppose), but as you mentioned yourself the result is pointless, so the success isn't desired anyway, and it would succeed just because Termux hasn't implemented versioned dependency.

> If files are moved between the packages and TERMUX_{SUB,}PKG_BREAKS isn't set however, then the command apt update && apt install pulseaudio would fail.

When versioned dependency isn't a thing, `apt update && apt install pulseaudio` isn't supposed to be run by users, because it doesn't upgrade libpulseaudio automatically; therefore we have no reason to make sure it doesn't fail, even when it doesn't fail when no files have been moved.
> Updating only pulseaudio would then of course give a none-working package since it's linked against a different library. But the actual command apt update && apt install pulseaudio would succeed.

I am not sure what's your point in this. Yes it would succeed (when no files are moved I suppose), but as you mentioned yourself the result is pointless, so the success isn't desired anyway, and it would succeed just because Termux hasn't implemented versioned dependency.

> If files are moved between the packages and TERMUX_{SUB,}PKG_BREAKS isn't set however, then the command apt update && apt install pulseaudio would fail.

When versioned dependency isn't a thing, `apt update && apt install pulseaudio` isn't supposed to be run by users, because it doesn't upgrade libpulseaudio automatically; therefore we have no reason to make sure it doesn't fail, even when it doesn't fail when no files have been moved.

**Comment by tomty89** [comment]
> Updating only pulseaudio would then of course give a none-working package since it's linked against a different library. But the actual command apt update && apt install pulseaudio would succeed.

I am not sure what's your point in this. Yes it would succeed (when no files are moved I suppose), but as you mentioned yourself the result is pointless, so the success isn't desired anyway, and it would succeed just because Termux hasn't implemented versioned dependency.

> If files are moved between the packages and TERMUX_{SUB,}PKG_BREAKS isn't set however, then the command apt update && apt install pulseaudio would fail.

When versioned dependency isn't a thing, `apt update && apt install pulseaudio` isn't supposed to be run by users, because it doesn't upgrade libpulseaudio automatically; therefore we have no reason to make sure it doesn't fail, even when it doesn't fail when no files have been moved.

**Comment by tomty89** [comment]
@Grimler91 I think the only concern now is whether `TERMUX_PKG_BREAKS="pulseaudio (<< 11.1-5)` (I withdrew it from my commit btw) alone is enough to make `apt update && apt install libpulseaudio pulseaudio` / `apt upgrade` work. It would be great if you can help to test that and report back the new `dpkg -L` on `{lib,lib,}pulseaudio{,-dev,}`. (I don't have access to a x86_64 device so I cannot test them myself).

File: packages/libpulseaudio/build.sh [hunk 1]
```diff
@@ -28,25 +28,38 @@ termux_step_pre_configure () {
...
 	# so disable hardware detection.
 	sed -i $TERMUX_PREFIX/etc/pulse/default.pa \
 		-e '/^load-module module-detect$/s/^/#/'
 	echo "load-module module-sles-sink" >> $TERMUX_PREFIX/etc/pulse/default.pa
-	cd $TERMUX_PREFIX/libexec
 
+	cd $TERMUX_PREFIX/bin
+	for bin in pamon paplay parec parecord; do
+		rm -f ../libexec/$bin
+		ln -s ../libexec/pacat $bin
```
Comment (Grimler91) [inline packages/libpulseaudio/build.sh]: @tomty89 suppose to be `ln -s ../libexec/$bin $bin` here?

Comment (tomty89) [reply packages/libpulseaudio/build.sh]: Uhm no, it was supposed to be `ln -s ../bin/pacat $bin`. They were symlinks to `pacat` in `libexec/`. But I just notice that the `$0` magic doesn't work with a wrapper. Going to redo this.

**Comment by Grimler91** [comment]
@tomty89 I opened a pull request to your branch.

**Comment by Grimler91** [comment]
Here are the file lists for the updated packages:

libpulseaudio:
```
/.
/data
/data/data
/data/data/com.termux
/data/data/com.termux/files
/data/data/com.termux/files/usr
/data/data/com.termux/files/usr/bin
/data/data/com.termux/files/usr/bin/esdcompat
/data/data/com.termux/files/usr/bin/pacat
/data/data/com.termux/files/usr/bin/pacmd
/data/data/com.termux/files/usr/bin/pactl
/data/data/com.termux/files/usr/bin/pasuspender
/data/data/com.termux/files/usr/bin/pulseaudio
/data/data/com.termux/files/usr/bin/pamon
/data/data/com.termux/files/usr/bin/paplay
/data/data/com.termux/files/usr/bin/parec
/data/data/com.termux/files/usr/bin/parecord
/data/data/com.termux/files/usr/etc
/data/data/com.termux/files/usr/etc/pulse
/data/data/com.termux/files/usr/etc/pulse/system.pa
/data/data/com.termux/files/usr/etc/pulse/daemon.conf
/data/data/com.termux/files/usr/etc/pulse/client.conf
/data/data/com.termux/files/usr/etc/pulse/default.pa
/data/data/com.termux/files/usr/lib
/data/data/com.termux/files/usr/lib/pulseaudio
/data/data/com.termux/files/usr/lib/pulseaudio/libpulsecommon-11.1.so
/data/data/com.termux/files/usr/lib/pulseaudio/libpulsecore-11.1.so
/data/data/com.termux/files/usr/lib/libpulse.so
/data/data/com.termux/files/usr/lib/libpulse-simple.so
/data/data/com.termux/files/usr/libexec
/data/data/com.termux/files/usr/libexec/pulseaudio
/data/data/com.termux/files/usr/libexec/pacat
/data/data/com.termux/files/usr/libexec/pactl
/data/data/com.termux/files/usr/libexec/pasuspender
/data/data/com.termux/files/usr/libexec/pacmd
/data/data/com.termux/files/usr/libexec/esdcompat
/data/data/com.termux/files/usr/share
/data/data/com.termux/files/usr/share/man
/data/data/com.termux/files/usr/share/man/man1
/data/data/com.termux/files/usr/share/man/man1/pulseaudio.1
/data/data/com.termux/files/usr/share/man/man1/esdcompat.1
/data/data/com.termux/files/usr/share/man/man1/pax11publish.1
/data/data/com.termux/files/usr/share/man/man1/pacat.1
/data/data/com.termux/files/usr/share/man/man1/pacmd.1
/data/data/com.termux/files/usr/share/man/man1/pactl.1
/data/data/com.termux/files/usr/share/man/man1/pasuspender.1
/data/data/com.termux/files/usr/share/man/man1/padsp.1
/data/data/com.termux/files/usr/share/man/man1/start-pulseaudio-x11.1
/data/data/com.termux/files/usr/share/man/man5
/data/data/com.termux/files/usr/share/man/man5/pulse-daemon.conf.5
/data/data/com.termux/files/usr/share/man/man5/pulse-client.conf.5
/data/data/com.termux/files/usr/share/man/man5/default.pa.5
/data/data/com.termux/files/usr/share/man/man5/pulse-cli-syntax.5
/data/data/com.termux/files/usr/lib/libpulsecommon-11.1.so
/data/data/com.termux/files/usr/lib/libpulsecore-11.1.so
/data/data/com.termux/files/usr/libexec/parec
/data/data/com.termux/files/usr/libexec/pamon
/data/data/com.termux/files/usr/libexec/paplay
/data/data/com.termux/files/usr/libexec/parecord
/data/data/com.termux/files/usr/share/man/man1/paplay.1
/data/data/com.termux/files/usr/share/man/man1/parec.1
/data/data/com.termux/files/usr/share/man/man1/parecord.1
/data/data/com.termux/files/usr/share/man/man1/pamon.1
```

pulseaudio:
```
/.
/data
/data/data
/data/data/com.termux
/data/data/com.termux/files
/data/data/com.termux/files/usr
/data/data/com.termux/files/usr/lib
/data/data/com.termux/files/usr/lib/pulse-11.1
/data/data/com.termux/files/usr/lib/pulse-11.1/modules
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/libcli.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/libprotocol-cli.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/libprotocol-simple.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/libprotocol-http.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/libprotocol-native.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/librtp.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-cli.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-cli-protocol-tcp.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-simple-protocol-tcp.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-null-sink.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-null-source.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-sine-source.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-detect.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-volume-restore.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-device-manager.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-device-restore.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-stream-restore.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-card-restore.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-default-device-restore.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-always-sink.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-rescue-streams.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-intended-roles.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-suspend-on-idle.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-echo-cancel.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-http-protocol-tcp.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-sine.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-native-protocol-tcp.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-native-protocol-fd.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-combine.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-combine-sink.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-remap-sink.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-remap-source.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-ladspa-sink.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-tunnel-sink-new.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-tunnel-source-new.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-tunnel-sink.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-tunnel-source.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-position-event-sounds.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-augment-properties.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-role-cork.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-loopback.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-virtual-sink.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-virtual-source.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-virtual-surround-sink.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-switch-on-connect.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-switch-on-port-available.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-filter-apply.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-filter-heuristics.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-role-ducking.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-allow-passthrough.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-rtp-send.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-rtp-recv.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-cli-protocol-unix.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-simple-protocol-unix.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-http-protocol-unix.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-native-protocol-unix.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-pipe-sink.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-pipe-source.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-match.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-mmkbd-evdev.so
/data/data/com.termux/files/usr/lib/pulse-11.1/modules/module-sles-sink.so
/data/data/com.termux/files/usr/share
/data/data/com.termux/files/usr/share/bash-completion
/data/data/com.termux/files/usr/share/bash-completion/completions
/data/data/com.termux/files/usr/share/bash-completion/completions/pulseaudio
/data/data/com.termux/files/usr/share/zsh
/data/data/com.termux/files/usr/share/zsh/site-functions
/data/data/com.termux/files/usr/share/zsh/site-functions/_pulseaudio
/data/data/com.termux/files/usr/lib/libcli.so
/data/data/com.termux/files/usr/lib/libprotocol-cli.so
/data/data/com.termux/files/usr/lib/libprotocol-http.so
/data/data/com.termux/files/usr/lib/libprotocol-native.so
/data/data/com.termux/files/usr/lib/libprotocol-simple.so
/data/data/com.termux/files/usr/lib/librtp.so
/data/data/com.termux/files/usr/share/bash-completion/completions/pactl
/data/data/com.termux/files/usr/share/bash-completion/completions/pacmd
/data/data/com.termux/files/usr/share/bash-completion/completions/pasuspender
/data/data/com.termux/files/usr/share/bash-completion/completions/padsp
/data/data/com.termux/files/usr/share/bash-completion/completions/pacat
/data/data/com.termux/files/usr/share/bash-completion/completions/paplay
/data/data/com.termux/files/usr/share/bash-completion/completions/parec
/data/data/com.termux/files/usr/share/bash-completion/completions/parecord
```

libpulseaudio-dev:
```
/.
/data
/data/data
/data/data/com.termux
/data/data/com.termux/files
/data/data/com.termux/files/usr
/data/data/com.termux/files/usr/include
/data/data/com.termux/files/usr/include/pulse
/data/data/com.termux/files/usr/include/pulse/cdecl.h
/data/data/com.termux/files/usr/include/pulse/channelmap.h
/data/data/com.termux/files/usr/include/pulse/context.h
/data/data/com.termux/files/usr/include/pulse/def.h
/data/data/com.termux/files/usr/include/pulse/direction.h
/data/data/com.termux/files/usr/include/pulse/error.h
/data/data/com.termux/files/usr/include/pulse/ext-device-manager.h
/data/data/com.termux/files/usr/include/pulse/ext-device-restore.h
/data/data/com.termux/files/usr/include/pulse/ext-stream-restore.h
/data/data/com.termux/files/usr/include/pulse/format.h
/data/data/com.termux/files/usr/include/pulse/gccmacro.h
/data/data/com.termux/files/usr/include/pulse/introspect.h
/data/data/com.termux/files/usr/include/pulse/mainloop-api.h
/data/data/com.termux/files/usr/include/pulse/mainloop-signal.h
/data/data/com.termux/files/usr/include/pulse/mainloop.h
/data/data/com.termux/files/usr/include/pulse/operation.h
/data/data/com.termux/files/usr/include/pulse/proplist.h
/data/data/com.termux/files/usr/include/pulse/pulseaudio.h
/data/data/com.termux/files/usr/include/pulse/rtclock.h
/data/data/com.termux/files/usr/include/pulse/sample.h
/data/data/com.termux/files/usr/include/pulse/scache.h
/data/data/com.termux/files/usr/include/pulse/simple.h
/data/data/com.termux/files/usr/include/pulse/stream.h
/data/data/com.termux/files/usr/include/pulse/subscribe.h
/data/data/com.termux/files/usr/include/pulse/thread-mainloop.h
/data/data/com.termux/files/usr/include/pulse/timeval.h
/data/data/com.termux/files/usr/include/pulse/utf8.h
/data/data/com.termux/files/usr/include/pulse/util.h
/data/data/com.termux/files/usr/include/pulse/version.h
/data/data/com.termux/files/usr/include/pulse/volume.h
/data/data/com.termux/files/usr/include/pulse/xmalloc.h
/data/data/com.termux/files/usr/share
/data/data/com.termux/files/usr/share/vala
/data/data/com.termux/files/usr/share/vala/vapi
/data/data/com.termux/files/usr/share/vala/vapi/libpulse.deps
/data/data/com.termux/files/usr/share/vala/vapi/libpulse.vapi
/data/data/com.termux/files/usr/share/vala/vapi/libpulse-mainloop-glib.deps
/data/data/com.termux/files/usr/share/vala/vapi/libpulse-mainloop-glib.vapi
/data/data/com.termux/files/usr/share/vala/vapi/libpulse-simple.deps
/data/data/com.termux/files/usr/share/vala/vapi/libpulse-simple.vapi
/data/data/com.termux/files/usr/lib
/data/data/com.termux/files/usr/lib/pkgconfig
/data/data/com.termux/files/usr/lib/pkgconfig/libpulse.pc
/data/data/com.termux/files/usr/lib/pkgconfig/libpulse-simple.pc
/data/data/com.termux/files/usr/lib/cmake
/data/data/com.termux/files/usr/lib/cmake/PulseAudio
/data/data/com.termux/files/usr/lib/cmake/PulseAudio/PulseAudioConfig.cmake
/data/data/com.termux/files/usr/lib/cmake/PulseAudio/PulseAudioConfigVersion.cmake
```

**Comment by tomty89** [comment]
Oops seems like `{}` wouldn't work. Guess I'll have to break them

**Comment by tomty89** [comment]
@Grimler91 Thanks! I added another commit to fix TERMUX_SUBPKG_INCLUDE.

**Pull request closed without merging**